components/provider_cldr/src/transform/dates.rs

```diff
@@ -140,7 +143,80 @@ impl From<&cldr_json::DateTimeFormats> for gregory::patterns::DateTimeFormatsV1
-            available_formats: other.available_formats.0.clone(),
+            skeletons: {
+                let mut skeletons: SkeletonsV1 = Vec::new();
+
+                // Matching up patterns to their unique skeletons is faster if the initial list
+                // of available_formats is sorted, while the data provided by the CLDR is not
+                // strictly sorted.
+                let mut sorted_available_formats = other.available_formats.0.clone();
+                sorted_available_formats.sort();
+
+                // The CLDR keys for available_formats can have duplicate skeletons with either
+                // an additional variant, or with multiple variants for different plurals.
+                for (skeleton_str, pattern_str) in sorted_available_formats.iter() {
+                    let mut unique_skeleton = String::new();
+                    let mut variant_text = String::new();
+
+                    // Split out the unique skeleton component, and the variant text.
+                    //   "MMMMW-count-two"
+                    //   unique_skeleton: "MMMMW"
+                    //   variant_text: "-count-two"
+                    let chars = skeleton_str.chars();
+                    let mut target = &mut unique_skeleton;
+                    for ch in chars {
+                        if ch == '-' {
+                            target = &mut variant_text;
+                        }
+                        target.push(ch);
+                    }
+
+                    let skeleton_pattern = SkeletonPatternV1 {
+                        pattern: pattern_str.clone(),
+                        count: match variant_text.as_str() {
+                            // Alt
+                            "" | "-alt-variant" => None,
+                            "-count-one" => Some(CountV1::One),
+                            "-count-other" => Some(CountV1::Other),
+                            "-count-zero" => Some(CountV1::Zero),
+                            "-count-two" => Some(CountV1::Two),
+                            "-count-few" => Some(CountV1::Few),
+                            "-count-many" => Some(CountV1::Many),
+                            variant => panic!("Unknown skeleton variant found {:?}", variant),
+                        },
+                    };
+
+                    let unique_skeleton = Cow::Owned(unique_skeleton);
+
+                    // The skeletons are sorted, so the variants like "-alt-variant" or "-count-many"
+                    // will always be next to the unique skeleton.
+                    let mut existing_patterns =
+                        skeletons
+                            .last_mut()
+                            .and_then(|(last_skeleton, last_patterns)| {
+                                debug_assert!(
+                                    *last_skeleton <= unique_skeleton,
+                                    "The skeletons must be sorted."
+                                );
```

Comment:
This is an assertion for the sort that I perform earlier, so if it fails it means there is a programming error, not a data error. I believe keeping it as a `debug_assert` is the proper choice here, as it will only fail if someone removes that sort. The data from CLDR is not correctly sorted alphabetically. I could make this more clear with an additional comment.

https://github.com/unicode-org/icu4x/pull/480/files/63983715c4af1caed5877d7f4a61ba66da7a4677..a5e20def6e06d671114dfe51ca831ecfbd148467#diff-4313b054e5b8d61c9fba3243afb5070dc55e8caae89d8922d8970cf1bd97ef3bR149-R153

Reply:
Thanks; didn't see the sort earlier.